design/secret-delegation.md

```diff
@@ -0,0 +1,84 @@
+# Secret delegation specficiation
+
+_Status_: Draft, in review.
+
+This document outlines a specification for delegating of permission to use the contents of a k8s Secret object from the owner's namespace to another namespace.
+
+## Goals
+
+- Permit wildcard certificates to be referenced across namespaces.
+
+## Non-goals
+
+- To implement a notion of a default, or fallback, TLS certificate available to all Ingress/IngressRoute objects.
+- To permit other classes of kubernetes objects to be referenced across namespaces.
+
+## Background
+
+Currently the Secret containing the TLS certificate must be co-located with the Ingress or root IngressRoute object referencing the secret.
+This requirement complicates deployment patterns where wildcard TLS certificates are used, specifically the use of a wildcard certificate to secure a number of subdomains.
+e.g. presenting foo.example.com using the certifiate for \*.example.com when the Ingress/IngressRoute object for foo.example.com does not share the same namespace as the Secret containing \*.example.com.
+
+This proposal introduces a specification, in the form of a simple CRD, whereby the permission to access a secret is delegated from the owning namespace to one or more other namespaces.
+
+## High-Level Design
+
+- The definition of a Secret's name is extended to recognize a namespace prefix, ie. `secretName: kube-system/wildcard-tls` means the secret `wildcard-tls` in the `kube-system` namespace.
+- A SecretDelegation CRD creates a mapping between a Secret in the owners namespace and the permission to reference that secret from a different namespace.
+
+## Detailed Design
+
+The implementation of this design is in three parts; the addition of a SecretDelegation CRD, and the modifications to the interpretation the tls stanza in Ingress and IngressRoute objects.
+
+### SecretDelegation object
+
+The SecretDelegation object provides a mapping of Secret objects from the SecretDelegation's namespace to the supplied target namespaces.
+
+```
+apiVersion: vmware.com/v1
+kind: SecretDelegation
+metadata:
+  name: wildcards
+  namespace: kube-system
+spec:
+  delegations:
+  - secretName: example-com-wildcard
+    targetNamespaces:
```

Comment:
It might be worth documenting how `targetNamespaces` interacts with how IngressRoute describes restricted delegation of particular vhosts.

If I have a wildcard certificate for `*.example.com`, I might intend to allow it to be referenced by the `foo` namespace, but _only_ for the `foo.example.com` vhost. I believe that latter restriction could be enforced with the existing IngressRoute mechanisms but it's not called out explicitly here.

Reply:
> It might be worth documenting how `targetNamespaces` interacts with how IngressRoute describes restricted delegation of particular vhosts.

I'd like that to be orthogonal. I have an alternative proposal that creates a coupling between wildcards and delegation.

> If I have a wildcard certificate for `*.example.com`, I might intend to allow it to be referenced by the `foo` namespace, but _only_ for the `foo.example.com` vhost. I believe that latter restriction could be enforced with the existing IngressRoute mechanisms but it's not called out explicitly here.

If you want to do that, and you don't trust the owner of foo.example.com with your certificate, then, the IngressRoute pattern of placing the root vhost for foo.example.com in one namespace, along with the certificate (a wildcard certificate can be used for a named subdomain without issue), and then delegating the responsibility for all the paths under foo.example.com to another namespace via IngressRoute's `prefix: "/"` delegation.